### PR TITLE
[Op][Core] Add ScatterNDUpdate-14 core and reference

### PR DESCRIPTION
**Pull request by mmikolajcz**
### Details:
 - *Add core implementation for ScatterNDUpdates-14*
 - *Add reference implementation for ScatterNDUpdates-14*

### Tickets:
 - *111098*
 - *111091*
 - *111093*

### CODE REVIEW

**Review by rkazants — CHANGES_REQUESTED** [review]
It looks not covering TF case with duplicating indices: https://www.tensorflow.org/api_docs/python/tf/raw_ops/TensorScatterUpdate. We initially agreed to add such support not only reduction modes
Will we support it?

**Comment by mmikolajcz** [comment]
> It looks not covering TF case with duplicating indices: https://www.tensorflow.org/api_docs/python/tf/raw_ops/TensorScatterUpdate. We initially agreed to add such support not only reduction modes Will we support it?

Yes, sorry, forgot to add test cases for that scenario: https://github.com/openvinotoolkit/openvino/pull/23754/files#diff-b3357cd694cf99989466fe2fb084606a048a1af52c233b36312e3aac6be2a6c8R257-R325
In case of indices duplicates in copy mode, only last update for given duplicated index will be used. Existing reference implementation seem to support this case.
In case you have some cases from TF that should be added to those test suites, feel free to contact me, I'll update them

**Comment by rkazants** [comment]
> > It looks not covering TF case with duplicating indices: https://www.tensorflow.org/api_docs/python/tf/raw_ops/TensorScatterUpdate. We initially agreed to add such support not only reduction modes Will we support it?
> 
> Yes, sorry, forgot to add test cases for that scenario: https://github.com/openvinotoolkit/openvino/pull/23754/files#diff-b3357cd694cf99989466fe2fb084606a048a1af52c233b36312e3aac6be2a6c8R257-R325 In case of indices duplicates in copy mode, only last update for given duplicated index will be used. Existing reference implementation seem to support this case. In case you have some cases from TF that should be added to those test suites, feel free to contact me, I'll update them

Cool, please also update specification with such a case. Should we add additional attribute to `ScatterNDUpdate` for such mode?
Please also share a link to updated specification.

Best regards,
Roman

File: src/core/reference/include/openvino/reference/scatter_nd_update.hpp [hunk 1]
```diff
@@ -7,19 +7,93 @@
 #include <cstring>
 #include <numeric>
 
+#include "add.hpp"
+#include "and.hpp"
+#include "maximum.hpp"
+#include "minimum.hpp"
+#include "multiply.hpp"
```
Comment (praasz) [inline src/core/reference/include/openvino/reference/scatter_nd_update.hpp:14]: Are these includes required?
Include them as "openvino/reference/add.hpp"

File: src/core/src/op/scatter_nd_update.cpp [hunk 2]
```diff
@@ -140,5 +144,120 @@ bool ScatterNDUpdate::evaluate_symbol(TensorSymbolVector& output_symbols) const
...
+                                      scatter_nd_update::Evaluate,
+                                      data.get_element_type(),
+                                      data,
+                                      indices,
+                                      updates,
+                                      output,
+                                      data_shape,
+                                      indices_shape,
+                                      updates_shape,
+                                      m_reduction);
```
Comment (praasz) [inline src/core/src/op/scatter_nd_update.cpp:197]: Looks same as for older version make common local helper to share the code.

File: src/core/src/op/scatter_nd_update.cpp [hunk 3]
```diff
@@ -140,5 +144,120 @@ bool ScatterNDUpdate::evaluate_symbol(TensorSymbolVector& output_symbols) const
...
+    default:
+        return false;
+    }
+    switch (get_input_element_type(1)) {
+    case element::i32:
+    case element::i64:
+        return true;
+    default:
+        return false;
+    }
```
Comment (praasz) [inline src/core/src/op/scatter_nd_update.cpp:221]: Looks same as for older version make local helper and share code (single helper or maybe separate for both each input).

File: src/core/tests/type_prop/scatter_nd_update.cpp [hunk 4]
```diff
@@ -109,84 +122,73 @@ TEST(type_prop, scatter_nd_update_fail_indices_last_dim) {
...
-protected:
-    void SetUp() override {
-        set_shape_symbols(data_3d_dynamic);
-    }
-    PartialShape data_3d_dynamic{{2, 5}, 2, {4, 10}};
-};
-
-TEST_F(TypePropScatterUpdateNDV3Test, data_input_partial_shape_and_symbols_propagation) {
-    const auto d = std::make_shared<Parameter>(element::f32, data_3d_dynamic);
+TYPED_TEST_P(TypePropScatterNDUpdateTest, data_input_partial_shape_and_symbols_propagation) {
```
Comment (praasz) [inline src/core/tests/type_prop/scatter_nd_update.cpp:125]: Add some dedicated test for new op to check new attribute is set correctly, has expected default value etc.

**Review by praasz — CHANGES_REQUESTED** [review]

File: src/core/tests/type_prop/scatter_nd_update.cpp (hunk 4 above)
Comment (mmikolajcz) [reply src/core/tests/type_prop/scatter_nd_update.cpp:124]: Added 3 additional attribute tests for default, using constructor and setter

File: src/core/reference/include/openvino/reference/scatter_nd_update.hpp [hunk 5]
```diff
@@ -8,18 +8,92 @@
 #include <numeric>
 
 #include "openvino/core/shape.hpp"
+#include "openvino/op/scatter_nd_update.hpp"
+#include "openvino/reference/maximum.hpp"
+#include "openvino/reference/minimum.hpp"
+#include "openvino/reference/multiply.hpp"
+#include "openvino/reference/xor.hpp"
+#include "subtract.hpp"
```
Comment (praasz) [inline src/core/reference/include/openvino/reference/scatter_nd_update.hpp:16]:
```suggestion
#include "openvino/reference/subtract.hpp"
```

**Review by praasz — APPROVED** [review]

File: src/core/reference/include/openvino/reference/scatter_nd_update.hpp (hunk 5 above)
Comment (mmikolajcz) [reply src/core/reference/include/openvino/reference/scatter_nd_update.hpp]: Thanks, missed this include, removed instead

**Review by mitruska — APPROVED** [review]
Just a reminder to register `shape_infer` for CPU (can be done as a follow up).
There is a common shape_infer for `ScatterNDBase` that can be reused for the new version of the op.
https://github.com/openvinotoolkit/openvino/blob/34825fe2c2ccda54f8b6c2afac3a724e730a526b/src/plugins/intel_cpu/src/shape_inference/shape_inference.cpp#L470

**Review by rkazants — APPROVED** [review]

**Pull request merged**